File: externals/Async-CTP/samples/VB/Excel/OptionPricing/ExcelInterop.vb
Imports System.Threading
Imports System.Runtime.CompilerServices

Public Module ExcelInterop

    <Extension()>
    Public Sub StartAsyncMethod(Application As Microsoft.Office.Interop.Excel.Application)
        If s_app Is Nothing Then s_app = Application
        If s_UI_SC Is Nothing Then
            If SynchronizationContext.Current Is Nothing Then
                Using c As New System.Windows.Forms.Control : Dim dummy = c.Handle : End Using
            End If
            s_UI_SC = SynchronizationContext.Current
        End If
        If s_Excel_SC Is Nothing Then s_Excel_SC = New ExcelSynchronizationContext
        If Not (TypeOf SynchronizationContext.Current Is ExcelSynchronizationContext) Then SynchronizationContext.SetSynchronizationContext(s_Excel_SC)
    End Sub

    Private s_queue As New Queue(Of Tuple(Of SendOrPostCallback, Object))
    Private s_app As Microsoft.Office.Interop.Excel.Application = Nothing
    Private s_UI_SC As SynchronizationContext = Nothing
    Private s_Excel_SC As New ExcelSynchronizationContext

    Private Class ExcelSynchronizationContext
        Inherits SynchronizationContext

        Public Overrides Sub Send(d As System.Threading.SendOrPostCallback, state As Object)
            Throw New NotSupportedException
        End Sub

        Public Overrides Sub Post(d As System.Threading.SendOrPostCallback, state As Object)
            SyncLock s_queue
                s_queue.Enqueue(Tuple.Create(d, state))
                If s_queue.Count = 1 Then s_UI_SC.Post(AddressOf OnThread, Nothing)
            End SyncLock
        End Sub

        Private Sub OnThread(state As Object)
            SynchronizationContext.SetSynchronizationContext(s_Excel_SC)
            Dim isReady = False
            Try
                Dim range = s_app.Sheets(1).Cells(1, 1)
                range.Value = range.Value
                If s_app.Ready Then isReady = True
            Catch ex As Runtime.InteropServices.COMException
                If ex.ErrorCode <> -2146827284 AndAlso ex.ErrorCode <> -2146777998 Then Throw
            End Try

            If Not isReady Then
                Call New Timer(Sub(self)
                                   DirectCast(self, IDisposable).Dispose()
                                   s_UI_SC.Post(AddressOf OnThread, Nothing)
                               End Sub).Change(250, Timeout.Infinite)
            Else
                Dim q2 As Queue(Of Tuple(Of SendOrPostCallback, Object))
                SyncLock s_queue
                    q2 = New Queue(Of Tuple(Of SendOrPostCallback, Object))(s_queue)
                    s_queue.Clear()
                End SyncLock
                For Each ds In q2
                    ds.Item1.Invoke(ds.Item2)
                Next
            End If
        End Sub
    End Class

End Module

File: externals/Async-CTP/samples/CS/Excel/OptionPricing/ExcelInterop.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Threading;
using Microsoft.Office.Interop.Excel;

public static class ExcelInterop
{
    public static void StartAsyncMethod(this Microsoft.Office.Interop.Excel.Application Application)
    {
        if (s_app == null) s_app = Application;
        if (s_uiSC == null) { if (SynchronizationContext.Current == null) { var c = new System.Windows.Forms.Control(); var _ = c.Handle; c.Dispose(); } s_uiSC = SynchronizationContext.Current; }
        if (s_excelSC == null) s_excelSC = new ExcelSynchronizationContext();
        if ((SynchronizationContext.Current as ExcelSynchronizationContext) == null) SynchronizationContext.SetSynchronizationContext(s_excelSC);
    }

    private static Queue<Tuple<SendOrPostCallback, object>> s_queue = new Queue<Tuple<SendOrPostCallback, object>>();
    private static Microsoft.Office.Interop.Excel.Application s_app = null;
    private static SynchronizationContext s_uiSC = null;
    private static SynchronizationContext s_excelSC = new ExcelSynchronizationContext();

    private class ExcelSynchronizationContext : SynchronizationContext
    {
        public override void Send(SendOrPostCallback d, object state) { throw new NotSupportedException(); }

        public override void Post(SendOrPostCallback d, object state)
        {
            lock (s_queue)
            {
                s_queue.Enqueue(Tuple.Create(d, state));
                if (s_queue.Count == 1) s_uiSC.Post(OnThread, null);
            }
        }

        private void OnThread(object state)
        {
            SynchronizationContext.SetSynchronizationContext(s_excelSC);
            bool isReady = false;
            try
            {
                var range = (Range)((dynamic)s_app.Sheets[1]).Cells[1, 1];
                range.Value = range.Value;
                if (s_app.Ready) isReady = true;
            }
            catch (COMException ex) { if (ex.ErrorCode != -2146827284 && ex.ErrorCode != -2146777998) throw; }

            if (!isReady)
            {
                new Timer(self =>
                {
                    ((IDisposable)self).Dispose();
                    s_uiSC.Post(OnThread, null);
                }).Change(250, Timeout.Infinite);
            }
            else
            {
                Queue<Tuple<SendOrPostCallback, object>> q2;
                lock (s_queue)
                {
                    q2 = new Queue<Tuple<SendOrPostCallback, object>>(s_queue);
                    s_queue.Clear();
                }
                foreach (var ds in q2) ds.Item1(ds.Item2);
            }
        }
    }
}